### DISCUSSION TITLE
DSL-like call count verification

### DISCUSSION THREAD

**Discussion by T45K** [mockk/mockk#1227] [Ideas]
currently, in MockK, we need to separate `verify` blocks to verify call counts of mocked methods.

for example:
```kotlin
class Hello {
  fun hello(name: String) = ...
  fun goodBye(name: String) = ...
}

val mock = mockk<Hello> {
  justRun { hello(any()) }
  justRun { goodBye(any()) }
}

mock.hello("Bob")

verify(exactly = 1) { mock.hello("Bob") }
verify(exactly = 0) {
  mock.hello("Alice")
  mock.goodBye(any())
}
verify(atLeast = 0, atMost = 5) { mock.hello(any()) }
```

this is a bit troublesome and not readable for me (and maybe for us?).

i'd like to propose to introduce new DSL to verify them easily like below:

```kotlin
verify {
  1 * { mock.hello("Bob") }
  0 * { mock.hello("Alice") }
  0 * { mock.goodBye(any()) }
  (0..5) * { mock.hello(any()) }
}
```

this DSL can be implemented like below

```kotlin
class MockKCountVerificationScope {
  operator fun Int.times(lambda: MockKVerificationScope.() -> Unit) {
    verify(exactly = this, verifyBlock = lambda)
  }

  operator fun IntRange.times(lambda: MockKVerificationScope.() -> Unit) {
    verify(atLeast = this.first, atMost = this.last, verifyBlock = lambda)
  }
}

fun verify(verifyCountBlock: MockKCountVerificationScope.() -> Unit) {
  MockKCountVerificationScope().verifyCountBlock()
}
```

such DSL is introduced into other famous JVM mocking libraries e.g., [Mockito-Kotlin](https://github.com/mockito/mockito-kotlin/blob/main/mockito-kotlin/src/main/kotlin/org/mockito/kotlin/VerifyScope.kt#L44) and [Spock](https://spockframework.org/spock/docs/2.3/spock_primer.html#_interactions)

thank you for reading!

**Comment by Raibaz** [comment]
Ah, that'd be cool!

I used Spock in the past and loved that DSL :)

Would you be able to put together a PR to add this feature? 🙏

**Reply by T45K** [reply]
thank you for checking this discussion!

> Would you be able to put together a PR to add this feature? 🙏

thank you for your suggestion!
yes, i'd really like to do so

**Comment by T45K** [comment]
@Raibaz 
hi, i created a draft PR to implement this!
https://github.com/mockk/mockk/pull/1234

i'd appreciate if you could confirm this and give me feedback 😄
